File: Java/DeepspaceJava/deepspace/src/deepspace/GameUniverse.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author paula
 */
public class GameUniverse {
    private static int WIN = 10;
    private int turns;
    private int currentStationIndex;
    private Dice dice;
    private GameStateController gameState;
    private boolean haveSpaceCity;
    private ArrayList<SpaceStation> spaceStations;
    private SpaceStation currentStation;
    private EnemyStarShip currentEnemy;
    
    
    public GameUniverse (){
      turns = 0;
      gameState = new GameStateController();
      currentStationIndex = -1;
      dice = new Dice();
      currentStation = null;
      spaceStations = new ArrayList<>();
      currentEnemy = null;
      haveSpaceCity = false;
    }

    public CombatResult combat (SpaceStation station, EnemyStarShip enemy){
        GameCharacter ch = dice.firstShot();
        boolean enemyWins;
        float fire;
        ShotResult result;
        CombatResult combatResult;

        if (ch == GameCharacter.ENEMYSTARSHIP){
            fire = enemy.fire();
            result = station.receiveShot(fire);

            if (result == ShotResult.RESIST){
                fire = station.fire();
                result = enemy.receiveShot(fire);
                enemyWins = (result == ShotResult.RESIST);
            }
            else
                enemyWins = true;
        }
        else{
            fire = station.fire();
            result = enemy.receiveShot(fire);
            enemyWins = (result==ShotResult.RESIST);
        }

        if (enemyWins){
            float s = station.getSpeed();
            boolean moves = dice.spaceStationMoves(s);

            if (!moves){
                Damage damage = enemy.getDamage();
                station.setPendingDamage(damage);
                combatResult = CombatResult.ENEMYWINS;
            }
            else{
                station.move();
                combatResult = CombatResult.STATIONESCAPES;
            }
        }
        else{
            Loot aLoot = enemy.getLoot();
            Transformation transf = station.setLoot(aLoot);
            combatResult = CombatResult.STATIONWINSANDCONVERTS;
            
            if (transf == Transformation.GETEFFICIENT)
                makeStationEfficient();
            else if (transf == Transformation.SPACECITY && !haveSpaceCity)
                createSpaceCity();
            else
                combatResult = CombatResult.STATIONWINS;
        }

        gameState.next(turns, spaceStations.size());

        return combatResult;
    }
    
    public CombatResult combat (){
        if ((getState() == GameState.BEFORECOMBAT) || (getState() == GameState.INIT))
            return combat(currentStation, currentEnemy);
        else
            return CombatResult.NOCOMBAT;
    }
    
    private void createSpaceCity (){
        ArrayList<SpaceStation> rest = new ArrayList<>();
            
        for (int i=0; i<spaceStations.size(); i++)
            if (spaceStations.get(i) != currentStation)
                rest.add(spaceStations.get(i));
            
        currentStation = new SpaceCity(currentStation, rest);
        spaceStations.set(currentStationIndex, currentStation);
        haveSpaceCity = true; 
    }
    
    public void discardHangar (){
        if (currentStation != null)
            currentStation.discardHangar();  
    }

    public void discardShieldBooster (int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardShieldBooster(i); 
    }

    public void discardShieldBoosterInHangar(int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardShieldBoosterInHangar(i);
    }

    public void discardWeapon (int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardWeapon(i);
    }

    public void discardWeaponInHangar (int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardWeaponInHangar(i);
    }

    public GameState getState (){
      return gameState.getState();
    }

    public GameUniverseToUI getUIversion (){
      return new GameUniverseToUI(currentStation, currentEnemy);
    }

    public boolean haveAWinner (){
        if (currentStation != null)
            return (currentStation.getNMedals() >= WIN);
        else
            return false;
    }

    public void init (ArrayList<String> names){
        if (getState() == GameState.CANNOTPLAY){
            CardDealer dealer = CardDealer.getInstance();

            for (int i=0; i< names.size(); i++){
                SuppliesPackage supplies = dealer.nextSuppliesPackage();
                SpaceStation station = new SpaceStation(names.get(i), supplies);
                spaceStations.add(station);
                int nh = dice.initWithNHangars();
                int nw = dice.initWithNWeapons();
                int ns = dice.initWithNShields();
                Loot lo = new Loot(0, nw, ns, nh, 0);
                station.setLoot(lo);
            }

            currentStationIndex = dice.whoStarts(names.size());
            currentStation = spaceStations.get(currentStationIndex);
            currentEnemy = dealer.nextEnemy();
            gameState.next(turns, spaceStations.size());
        }
    }
    
    private void makeStationEfficient (){
        if (dice.extraEfficiency())
            currentStation = new BetaPowerEfficientSpaceStation(currentStation);
        else
            currentStation = new PowerEfficientSpaceStation(currentStation);
        
        spaceStations.set(currentStationIndex, currentStation);
    }

    public void mountShieldBooster (int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.mountShieldBooster(i);
    }

    public void mountWeapon (int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.mountWeapon(i);  
    }

    public boolean nextTurn (){
        if (getState() == GameState.AFTERCOMBAT){
            boolean stationState = currentStation.validState();
            
            if (stationState){
                currentStationIndex = (currentStationIndex + 1) % spaceStations.size();
                turns += 1;
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedItems();
                CardDealer dealer = CardDealer.getInstance();
                currentEnemy = dealer.nextEnemy();
                gameState.next(turns, spaceStations.size());
                
                return true;
            }
            
            return false;
        }
        
        return false;
    }
}

File: Ruby/DeepSpaceRuby/lib/GameUniverse.rb
#encoding:utf-8

require_relative 'GameUniverseToUI'
require_relative 'Dice'
require_relative 'GameStateController'
require_relative 'CardDealer'
require_relative 'CombatResult'
require_relative 'SpaceStation'
require_relative 'EnemyStarShip'
require_relative 'PowerEfficientSpaceStation'
require_relative 'BetaPowerEfficientSpaceStation'
require_relative 'SpaceCity'

module Deepspace
  class GameUniverse
    @@WIN = 10

    def initialize
      @turns = 0
      @gameState = GameStateController.new
      @currentStationIndex = -1
      @dice = Dice.new
      @currentStation = nil
      @spaceStations = Array.new
      @currentEnemy = nil
      @haveSpaceCity = false
    end

    public
    def combatGo (station, enemy)
      ch = @dice.firstShot

      if ch == GameCharacter::ENEMYSTARSHIP then
        fire = enemy.fire
        result = station.receiveShot(fire)

        if result == ShotResult::RESIST then
          fire = station.fire
          result = enemy.receiveShot(fire)
          enemyWins = (result==ShotResult::RESIST)
        else
          enemyWins = true
        end
      else
        fire = station.fire
        result = enemy.receiveShot(fire)
        enemyWins = (result==ShotResult::RESIST)
      end

      if enemyWins then
        s = station.speed
        moves = @dice.spaceStationMoves(s)

        if !moves then
          damage = enemy.damage
          station.setPendingDamage(damage)
          combatResult = CombatResult::ENEMYWINS
        else
          station.move
          combatResult = CombatResult::STATIONESCAPES
        end
      else
        aLoot = enemy.loot
        transf = station.setLoot(aLoot)
        combatResult = CombatResult::STATIONWINSANDCONVERTS

        if transf == Transformation::GETEFFICIENT then
          makeStationEfficient
        elsif transf == Transformation::SPACECITY && !@haveSpaceCity then
          createSpaceCity
        else
          combatResult = CombatResult::STATIONWINS
        end
      end

      @gameState.next(@turns, @spaceStations.size())

      combatResult
    end

    def combat
      if ((state == GameState::BEFORECOMBAT) || (state == GameState::INIT)) then
        combatGo(@currentStation, @currentEnemy)
      else
         CombatResult::NOCOMBAT
      end
    end

    def createSpaceCity
      if !@haveSpaceCity then
        rest = Array.new

        @spaceStations.each{ |s|
          if s != @currentStation then
            rest.push(s)
          end
        }

        @currentStation = SpaceCity.new(@currentStation, rest)
        @spaceStations[@currentStationIndex] = @currentStation
        @haveSpaceCity = true
      end
    end

    def discardHangar
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT
        @currentStation.discardHangar
      end
    end

    def discardShieldBooster (i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT then
        @currentStation.discardShieldBooster(i)
      end
    end

    def discardShieldBoosterInHangar(i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT then
        @currentStation.discardShieldBoosterInHangar(i)
      end
    end

    def discardWeapon (i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT then
        @currentStation.discardWeapon(i)
      end
    end

    def discardWeaponInHangar (i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT then
        @currentStation.discardWeaponInHangar(i)
      end
    end

    def state
      return @gameState.state
    end

    def getUIversion
      return GameUniverseToUI.new(@currentStation, @currentEnemy)
    end

    def haveAWinner
      if @currentStation != nil then
        return (@currentStation.nMedals >= @@WIN)
      else
        return false
      end
    end

    def init (names)
      if state == GameState::CANNOTPLAY then
        dealer = CardDealer.instance

        for i in 0..(names.length-1)
          supplies = dealer.nextSuppliesPackage
          station = SpaceStation.new(names[i], supplies.ammoPower, supplies.fuelUnits, supplies.shieldPower)
          @spaceStations.push(station)
          nh = @dice.initWithNHangars
          nw = @dice.initWithNWeapons
          ns = @dice.initWithNShields
          lo = Loot.new(0, nw, ns, nh, 0)
          station.setLoot(lo)
        end

        @currentStationIndex = @dice.whoStarts(names.size)
        @currentStation = @spaceStations[@currentStationIndex]
        @currentEnemy = dealer.nextEnemy
        @gameState.next(@turns, @spaceStations.size)
      end
    end

    def makeStationEfficient
      if @dice.extraEfficiency then
        @currentStation = BetaPowerEfficientSpaceStation.new(@currentStation)
      else
        @currentStation = PowerEfficientSpaceStation.new(@currentStation)
      end

      @spaceStations[@currentStationIndex] = @currentStation
    end

    def mountShieldBooster (i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT then
        @currentStation.mountShieldBooster(i)
      end
    end

    def mountWeapon (i)
      if @gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT then
        @currentStation.mountWeapon(i)
      end
    end

    def nextTurn
      if state == GameState::AFTERCOMBAT then
        stationState = @currentStation.validState

        if stationState then
          @currentStationIndex = (@currentStationIndex + 1) % @spaceStations.size
          @turns += 1
          @currentStation = @spaceStations[@currentStationIndex]
          @currentStation.cleanUpMountedItems
          dealer = CardDealer.instance
          @currentEnemy = dealer.nextEnemy
          @gameState.next(@turns, @spaceStations.size)

          return true
        end

        return false

      end

      return false
    end

    def to_s
      cad = "tuns: #{@turns}, gameState: #{@gameState.state.to_s}, currentStationIndex: #{@currentStationIndex}"
      cad +=  "dice: #{@dice.to_s}, currentStation: #{@currentStation.to_s}, currentEnemy: #{@currentEnemy}"
      cad += "\nOtras estaciones: \n"

      @spaceStations.each{ |sStation|
        if sStation != @currentStation
          cad += sStation.to_s
        end
      }

      return cad
    end

  end
end
